price/src/main.rs

```diff
@@ -0,0 +1,119 @@
+
+    #[clap(subcommand)]
+    cmd: Cmd,
+}
+
+impl Cli {
+    pub async fn run(self) -> Result<()> {
+        let settings = Settings::new(self.config)?;
+        self.cmd.run(settings).await
+    }
+}
+
+#[derive(Debug, clap::Subcommand)]
+pub enum Cmd {
+    Server(Server),
+}
+
+impl Cmd {
+    pub async fn run(&self, settings: Settings) -> Result<()> {
+        match self {
+            Self::Server(cmd) => cmd.run(&settings).await,
+        }
+    }
+}
+
+#[derive(Debug, clap::Args)]
+pub struct Server {}
+
+impl Server {
+    pub async fn run(&self, settings: &Settings) -> Result<()> {
+        tracing_subscriber::registry()
+            .with(tracing_subscriber::EnvFilter::new(&settings.log))
+            .with(tracing_subscriber::fmt::layer())
+            .init();
+
+        // Install the prometheus metrics exporter
+        poc_metrics::start_metrics(&settings.metrics)?;
+
+        // configure shutdown trigger
+        let (shutdown_trigger, shutdown) = triggered::trigger();
+        tokio::spawn(async move {
+            let _ = signal::ctrl_c().await;
+            shutdown_trigger.trigger()
+        });
+
+        // Initialize uploader
+        let (file_upload_tx, file_upload_rx) = file_upload::message_channel();
+        let file_upload =
+            file_upload::FileUpload::from_settings(&settings.output, file_upload_rx).await?;
+
+        let store_base_path = path::Path::new(&settings.cache);
+
+        // price generators
+        let mut hnt_price_generator =
+            PriceGenerator::new(settings, BlockchainTokenTypeV1::Hnt).await?;
+        let mut mobile_price_generator =
+            PriceGenerator::new(settings, BlockchainTokenTypeV1::Mobile).await?;
+        let mut iot_price_generator =
+            PriceGenerator::new(settings, BlockchainTokenTypeV1::Iot).await?;
+        let mut hst_price_generator =
+            PriceGenerator::new(settings, BlockchainTokenTypeV1::Hst).await?;
+
+        let (price_sink, mut price_sink_server) = file_sink::FileSinkBuilder::new(
+            FileType::PriceReport,
+            store_base_path,
+            concat!(env!("CARGO_PKG_NAME"), "_report_submission"),
+        )
+        .deposits(Some(file_upload_tx.clone()))
+        .roll_time(Duration::minutes(PRICE_SINK_ROLL_MINS))
+        .create()
+        .await?;
+
+        tokio::try_join!(
+            hnt_price_generator
+                .run(price_sink.clone(), &shutdown, BlockchainTokenTypeV1::Hnt)
+                .map_err(Error::from),
+            mobile_price_generator
+                .run(price_sink.clone(), &shutdown, BlockchainTokenTypeV1::Mobile)
+                .map_err(Error::from),
+            iot_price_generator
+                .run(price_sink.clone(), &shutdown, BlockchainTokenTypeV1::Iot)
+                .map_err(Error::from),
+            hst_price_generator
+                .run(price_sink, &shutdown, BlockchainTokenTypeV1::Hst)
+                .map_err(Error::from),
```

Comment:
since these generators are just going to be reaching a `None` else case when they check for a pricing account key do we even want to run them before those tokens have prices to be fetched? I would see running them if we were going to default to writing out a $0 pricing report but since we're not it seems like we could skip them.

for that matter, and since we are _only_ accessing these prices via the written reports would it be easier to have a single `PriceGenerator` struct that loops over all "active" tokens based on configuration, stores _only_ the last price in a map in its state and spawns a thread with a join handle for each active token, passing the previous price to the thread and A) checking for a newer price, B) writing out the report to the file sink with the current timestamp and C) returning either the previous or newer price via the join handle to the PriceGenerator?

then you only need to grab current time once and it doesn't need to be mutable, you simplify the state of the PriceGenerator process, there's only one continuous thread you need to worry about shutting down and it's just a matter of adding a token string to a config list in Settings to start checking for additional prices

Reply:
Yeah we thought about having a single PriceGenerator, but it doesn't really matter if the timestamps for different token types (as they become available) within the reports are the same as long as they are there. I believe it's already pretty simple as is imo and the state is minimal.

Reply:
Does it matter if one  price gets successfully updated and others do not? ie is there any intra dependencies on the prices being current on the consumer side ? If iot gets updated but for whatever reason hnt fails to update and continues failing for a while and whilst there may be steep changes in the hnt price

Reply:
That's a good question but I'd say not something we have to solve for first pass at least. So in the scenario where HNT updated and IOT failed, we'll presumably continue to serve the same data for IOT till we:

a) either have some redundancy with our RPC endpoints (assuming it was an RPC failure to fetch IOT in the above example)
b) or we kind of "derive" IOT price from HNT in some way

@madninja What are your thoughts on this? I could look into adding support for multiple RPC endpoints I suppose. And _maybe_ do a mean of the prices fetched from multiple RPC endpoints?